org/mozilla/jss/provider/java/security/RSAPSSAlgorithmParameters.java

```diff
@@ -0,0 +1,244 @@
+ *    ...  -- Allows for future expansion --
+ *  }
+ */
+public class RSAPSSAlgorithmParameters extends AlgorithmParametersSpi {
+    public final static AlgorithmId defaultHashAlg = new AlgorithmId(AlgorithmId.SHA_oid);
+    public final static AlgorithmId defaultMaskGenFunc  = new AlgorithmId(AlgorithmId.MGF1_oid);
+    public final static BigInt          defaultSaltLen = new BigInt(20);
+    public final static BigInt          defaultTrailerField = new BigInt(1);
+
+    private PSSParameterSpec spec = PSSParameterSpec.DEFAULT;
+    private AlgorithmId hashAlg = defaultHashAlg;
+    private AlgorithmId maskGenFunc = defaultMaskGenFunc;
+    private BigInt saltLen = defaultSaltLen;
+    private BigInt trailerField = defaultTrailerField;
+
+    public RSAPSSAlgorithmParameters() {}
+
+    @Override
+    protected void engineInit(AlgorithmParameterSpec paramSpec)
+            throws InvalidParameterSpecException {
+        spec = (PSSParameterSpec) paramSpec;
+        populateFromSpec();
+    }
+
+    @Override
+    protected <T extends AlgorithmParameterSpec> T engineGetParameterSpec(Class<T> paramSpec)
+            throws InvalidParameterSpecException  {
+        if (paramSpec.isAssignableFrom(PSSParameterSpec.class)) {
+            return paramSpec.cast(spec);
+        }
+
+        throw new InvalidParameterSpecException("Unknown parameter spec passed to PSS parameters object: " + paramSpec.getName());
+    }
+
+    @Override
+    protected void engineInit(byte[] params) throws IOException {
+        decode(new DerInputStream(params), params);
+    }
+
+    @Override
+    protected void engineInit(byte[] params, String format) throws IOException {
+        // Assume Der for now.
+        Assert.notReached("engineInit(byte[],String) not supported");
+        throw new IOException("engineInit(byte[],String) not supported");
+    }
+
+    @Override
+    protected byte[] engineGetEncoded() throws IOException {
+        DerOutputStream out = new DerOutputStream();
+        encode(out);
+        return out.toByteArray();
+
+    }
+
+    @Override
+    protected byte[] engineGetEncoded(String format) throws IOException {
+        //Assume Der for now.
+        Assert.notReached("engineGetEncoded(String format)) not supported");
+        throw new IOException("engineGetEncoded(String format)) not supported");
+    }
+
+    @Override
+    protected String engineToString() {
+        String str = new String("Mozilla-JSS PSSAlgorithmParameters " +  getClass().getName() + " HashAlg: " + spec.getDigestAlgorithm() + " MaskGenAlg: " + spec.getMGFAlgorithm() );
+        return str;
+    }
+
+    private void decode(DerInputStream in , byte[] encoded) throws IOException {
+        if (in == null) {
+            throw new IOException("Invalid input: got null DerInputStream");
+        }
+
+        // Sequence has 3 members, trailer field ignored
+        DerValue seq[] = in.getSequence(3);
+        if (seq.length != 3) {
```

Comment:
@jmagne So I'm confused here. The comment above says "trailer field ignored". I'm not entirely clear on how `getSequence` works, but if there was a trailer field, wouldn't it return an array of 4 `DerValue` elements? 

In particular `readVector(...)` uses `startLen` as the initialize size of the vector, but it'll grow if there are four fields.

Shouldn't this read `if (seq.length < 3 || seq.length > 4)`?